libs/custom-scanner/src/custom_scanner.ts

```diff
@@ -0,0 +1,747 @@
+    return result;
+  }
+
+  /**
+   * Waits for the scanner status to match the given predicate.
+   */
+  async waitForStatus(
+    predicate: (status: ScannerStatus) => boolean,
+    { timeout }: { timeout?: number } = {}
+  ): Promise<Result<Optional<ScannerStatus>, ErrorCode>> {
+    debug('waiting for status');
+    const statusWatcher = this.watchStatus();
+
+    if (typeof timeout === 'number') {
+      void sleep(timeout).then(() => {
+        debug('status watcher timed out');
+        statusWatcher.stop();
+      });
+    }
+
+    for await (const statusResult of statusWatcher) {
+      if (statusResult.isErr()) {
+        debug('status watcher error: %o', statusResult.err());
+        statusWatcher.stop();
+        return statusResult;
+      }
+
+      const status = statusResult.ok();
+      debug('status watcher status: %o', status);
+
+      if (predicate(status)) {
+        debug('status watcher predicate matched');
+        statusWatcher.stop();
+        return statusResult;
+      }
+    }
+
+    return ok(undefined);
+  }
+
+  /**
+   * Scans a sheet of paper and returns the resulting images. Note that this
+   * method will always return a pair of `ImageFromScanner` objects, even if the
+   * scanner only scanned one side of the paper. The image buffer for the side
+   * that was not scanned will be empty.
+   */
+  async scan(
+    scanParameters: ScanParameters
+  ): Promise<Result<SheetOf<ImageFromScanner>, ErrorCode>> {
+    const timer = time(debug, 'scan');
+    debug('scanning');
+
+    const { wantedScanSide: scanSide } = scanParameters;
+    const scannerImageA: ImageFromScanner = {
+      imageBuffer: Buffer.alloc(0),
+      imageWidth: 0,
+      imageHeight: 0,
+      imageDepth: scanParameters.imageColorDepth,
+      imageFormat: ImageFileFormat.Jpeg,
+      scanSide: ScanSide.A,
+      imageResolution: scanParameters.resolution,
+    };
+    const scannerImageB: ImageFromScanner = {
+      imageBuffer: Buffer.alloc(0),
+      imageWidth: 0,
+      imageHeight: 0,
+      imageDepth: scanParameters.imageColorDepth,
+      imageFormat: ImageFileFormat.Jpeg,
+      scanSide: ScanSide.B,
+      imageResolution: scanParameters.resolution,
+    };
+
+    timer.checkpoint('set scan parameters');
+    let setScanParametersResult = await this.setScanParametersInternal(
+      scanParameters
+    );
+
+    if (setScanParametersResult.isErr()) {
+      if (setScanParametersResult.err() === ErrorCode.JobNotValid) {
+        debug('job not valid, recreating job');
+        const createJobResult = await this.createJobInternal();
+
+        if (createJobResult.isErr()) {
+          return createJobResult;
+        }
+
+        setScanParametersResult = await this.setScanParametersInternal(
+          scanParameters
+        );
+      }
+
+      if (setScanParametersResult.isErr()) {
+        return setScanParametersResult;
+      }
+    }
+
+    const { unlock } = await this.publicApiMutex.asyncLock();
```

Comment:
Why do this here are not at the top? Just curious.

Reply:
No good reason. I'll move it.